Port conflict handling

TomRHandcock opened the pull request:
- Implements detection of port conflicts via socket binding upon attempting to start Mockzilla server.
  - Upon a conflict, throws an exception

Will keep #246 open upon merging this as I would like to wrap the `PortConflictException` thrown here in a Dart exception.

TomRHandcock committed:
feat(common): Implement `PortConflictException`

TomRHandcock committed:
feat(kmm): Implement `SocketIo`

Implement `SocketIo` to determine if an arbitrary port is available.

TomRHandcock committed:
feat(kmm): Add `SocketIo` to dependency injection

TomRHandcock committed:
feat(kmm): Throw custom exception upon port conflict

Upon calling `startServer()`, throw a `PortConflictException` if the desired port is already in use.

TomRHandcock committed:
style(kmm): Rename `SocketIoUtils` -> `MockSelectorManager`

Fix spotless issues

SamC-Apadmi commented on mockzilla/src/commonMain/kotlin/com/apadmi/mockzilla/lib/internal/utils/SocketIo.kt:
```diff
@@ -0,0 +1,11 @@
+    suspend fun isPortAvailable(port: Int): Boolean = runCatching {
+        val serverSocket = aSocket(selectorManager).tcp().bind("127.0.0.1", port)
+        serverSocket.dispose()
+    }.isSuccess
```
If given 0 as the port ktor will automatically bind it, not sure if that'll break this code? Might be worth handling 0 as a special case?

TomRHandcock replied:
@SamC-Apadmi Didn't realise, what would be expected in this special case? We allow Mockzilla to start or fail?

TomRHandcock replied:
Upon reviewing the docs, I see that ktor doesn't bind _to port 0_, it merely just treats port 0 as a special flag to select a random available port. I think in this case, it would be expected that we simply skip the check [as it is invalid] and allow the server to start.

TomRHandcock committed:
refactor(kmm): Refactor `SocketIo` dependencies

Provider `SocketBinder` interface to `SocketIo` instead of `SelectorManager` to allow for mocking in unit tests.

TomRHandcock committed:
fix(kmm): Update `SocketIo` for special handling of port 0

Ktor treats port 0 as a flag to use a random, available port. As such, `SocketIo.isPortAvailable` should always return `true` for this port.
